// @nosana/cli/dist/src/providers/modules/db/index.js
import fs from 'fs';
import os from 'os';
import { JSONFileSyncPreset } from 'lowdb/node';
const initial_state = {
    resources: {
        images: {},
        volumes: {},
    },
    flows: {},
    info: {
        country: '',
        network: {
            ip: '',
            ping_ms: 0,
            download_mbps: 0,
            upload_mbps: 0,
        },
        cpu: {
            model: '',
            physical_cores: 0,
            logical_cores: 0,
        },
        disk_gb: 0,
        ram_mb: 0,
        gpus: {
            devices: [],
            runtime_version: '0.0',
            cuda_driver_version: '0.0',
            nvml_driver_version: '0.0.0',
        },
    },
};
export class DB {
    constructor(configLocation) {
        if (configLocation[0] === '~') {
            configLocation = configLocation.replace('~', os.homedir());
        }
        fs.mkdirSync(configLocation, { recursive: true });
        this.db = JSONFileSyncPreset(`${configLocation}/nosana_db.json`, initial_state);
        if (!this.db.data.resources) {
            this.db.data.resources = initial_state.resources;
        }
        if (!this.db.data.flows) {
            this.db.data.flows = initial_state.flows;
        }
        if (!this.db.data.info) {
            this.db.data.info = initial_state.info;
        }
        this.db.write();
    }
}
