// @nosana/cli/dist/src/services/state/node/types.js
export const NODE_STATE_NAME = {
    NODE_STARTING: 'NODE_STARTING',
    NODE_STARTED: 'NODE_STARTED',
    PROVIDER_HEALTH_CHECKING: 'PROVIDER_HEALTH_CHECKING',
    PROVIDER_HEALTH_PASSED: 'PROVIDER_HEALTH_PASSED',
    PROVIDER_HEALTH_FAILED: 'PROVIDER_HEALTH_FAILED',
    API_SERVER_STARTING: 'API_SERVER_STARTING',
    API_SERVER_STARTED: 'API_SERVER_STARTED',
    API_SERVER_FAILED: 'API_SERVER_FAILED',
    HEALTH_CHECK_STARTING: 'HEALTH_CHECK_STARTING',
    HEALTH_CHECK_RUNNING: 'HEALTH_CHECK_RUNNING',
    HEALTH_CHECK_PASSED: 'HEALTH_CHECK_PASSED',
    HEALTH_CHECK_FAILED: 'HEALTH_CHECK_FAILED',
    BENCHMARK_STARTING: 'BENCHMARK_STARTING',
    BENCHMARK_RUNNING: 'BENCHMARK_RUNNING',
    BENCHMARK_PASSED: 'BENCHMARK_PASSED',
    BENCHMARK_FAILED: 'BENCHMARK_FAILED',
    RETRIVING_MARKET: 'RETRIVING_MARKET',
    RETRIVING_MARKET_PASSED: 'RETRIVING_MARKET_PASSED',
    RETRIVING_MARKET_FAILED: 'RETRIVING_MARKET_FAILED',
    JOINING_QUEUE: 'JOINING_QUEUE',
    JOINED_QUEUE: 'JOINED_QUEUE',
    JOINING_QUEUE_PASSED: 'JOINING_QUEUE_PASSED',
    JOINING_QUEUE_FAILED: 'JOINING_QUEUE_FAILED',
    JOB_STARTING: 'JOB_STARTING',
    JOB_STARTING_PASSED: 'JOB_STARTING_PASSED',
    JOB_STARTING_FAILED: 'JOB_STARTING_FAILED',
    JOB_RUNNING: 'JOB_RUNNING',
    JOB_COMPLETED: 'JOB_COMPLETED',
    JOB_STOPPED: 'JOB_STOPPED',
    JOB_FAILED: 'JOB_FAILED',
    IDLE: 'IDLE',
    UPDATING: 'UPDATING',
    RESTARTING: 'RESTARTING',
    SHUTTING_DOWN: 'SHUTTING_DOWN',
    NONE: 'NONE',
};
