// @nosana/cli/dist/src/services/state/job/types.js
export const JOB_STATE_NAME = {
    PULLING_IMAGE: 'PULLING_IMAGE',
    PULLING_IMAGE_SUCCESS: 'PULLING_IMAGE_SUCCESS',
    PULLING_IMAGE_FAILED: 'PULLING_IMAGE_FAILED',
    CREATING_VOLUME: 'CREATING_VOLUME',
    CREATING_VOLUME_SUCCESS: 'CREATING_VOLUME_SUCCESS',
    CREATING_VOLUME_FAILED: 'CREATING_VOLUME_FAILED',
    CREATING_NETWORK: 'CREATING_NETWORK',
    CREATING_NETWORK_SUCCESS: 'CREATING_NETWORK_SUCCESS',
    CREATING_NETWORK_FAILED: 'CREATING_NETWORK_FAILED',
    GET_CONTAINER: 'GET_CONTAINER',
    GET_CONTAINER_FAILED: 'GET_CONTAINER_FAILED',
    GET_CONTAINER_PASSED: 'GET_CONTAINER_PASSED',
    CONTAINER_STARTING: 'CONTAINER_STARTING',
    CONTAINER_FAILED: 'CONTAINER_FAILED',
    CONTAINER_STARTED: 'CONTAINER_STARTED',
    EXPOSED_URL_STARTED: 'EXPOSED_URL_STARTED',
    RUN_CONTAINER_OPERATION: 'RUN_CONTAINER_OPERATION',
    FINISH_CONTAINER_OPERATION: 'FINISH_CONTAINER_OPERATION',
    CONTAINER_OPERATION_FAILED: 'CONTAINER_OPERATION_FAILED',
    RETREIVING_JOB_DEFINATION: 'RETREIVING_JOB_DEFINATION',
    RETREIVED_JOB_DEFINATION: 'RETREIVED_JOB_DEFINATION',
    RETREIVING_JOB_DEFINATION_FAILED: 'RETREIVING_JOB_DEFINATION_FAILED',
    JOB_DEFINATION_VALIDATION: 'JOB_DEFINATION_VALIDATION',
    JOB_DEFINATION_VALIDATION_PASSED: 'JOB_DEFINATION_VALIDATION_PASSED',
    STARTING_NEW_FLOW: 'STARTING_NEW_FLOW',
    STARTED_NEW_FLOW: 'STARTED_NEW_FLOW',
    FLOW_FINISHED: 'FLOW_FINISHED',
    CONTINUE_EXISTING_FLOW: 'CONTINUE_EXISTING_FLOW',
    WAITING_FOR_JOB_TO_COMPLETE: 'WAITING_FOR_JOB_TO_COMPLETE',
    OPERATION_STARTING: 'OPERATION_STARTING',
    OPERATION_STARTED: 'OPERATION_STARTED',
    OPERATION_FAILED: 'OPERATION_FAILED',
    OPERATION_PASSED: 'OPERATION_PASSED',
    NONE: 'NONE',
};
